File: peoplecare-remote-camera/apps/remote_camera/android/app/src/main/kotlin/tv/peoplecare/remotecamera/ReconnectPolicy.kt
package tv.peoplecare.remotecamera

import kotlin.math.min
import kotlin.math.pow
import kotlin.math.roundToLong

/**
 * Exponential backoff with jitter. Same parameters as the Dart control
 * client and the Control Room (1 s, x2, max 30 s, jitter 50%).
 */
class ReconnectPolicy(
    private val initialMs: Long = 1_000,
    private val maxMs: Long = 30_000,
    private val multiplier: Double = 2.0,
    private val jitter: Double = 0.5,
    private val random: () -> Double = { Math.random() },
) {
    var attempts: Int = 0
        private set

    fun nextDelayMs(): Long {
        val base = min(maxMs.toDouble(), initialMs * multiplier.pow(attempts.toDouble()))
        attempts++
        return (base - base * jitter * random()).roundToLong()
    }

    fun reset() {
        attempts = 0
    }
}

File: peoplecare-remote-camera/apps/remote_camera/lib/src/control/reconnect_policy.dart
import 'dart:math';

/// Exponential backoff with jitter (1 s, x2, max 30 s, 50% jitter).
/// Same policy as the native SRT reconnection and the Control Room socket.
class ReconnectPolicy {
  ReconnectPolicy({
    this.initial = const Duration(seconds: 1),
    this.max = const Duration(seconds: 30),
    this.multiplier = 2.0,
    this.jitter = 0.5,
    Random? random,
  }) : _random = random ?? Random();

  final Duration initial;
  final Duration max;
  final double multiplier;
  final double jitter;
  final Random _random;
  int _attempt = 0;

  int get attempts => _attempt;

  Duration nextDelay() {
    final baseMs = min(max.inMilliseconds.toDouble(), initial.inMilliseconds * pow(multiplier, _attempt));
    _attempt++;
    final spread = baseMs * jitter * _random.nextDouble();
    return Duration(milliseconds: (baseMs - spread).round());
  }

  void reset() => _attempt = 0;
}
